prometheus/registry_test.go

```diff
@@ -746,38 +746,153 @@ func BenchmarkHandler(b *testing.B) {
 }
 
 func TestAlreadyRegistered(t *testing.T) {
-	reg := prometheus.NewRegistry()
 	original := prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Name:        "test",
+			Help:        "help",
+			ConstLabels: prometheus.Labels{"const": "label"},
+		},
+		[]string{"foo", "bar"},
+	)
+	equalButNotSame := prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Name:        "test",
+			Help:        "help",
+			ConstLabels: prometheus.Labels{"const": "label"},
+		},
+		[]string{"foo", "bar"},
+	)
+	originalWithoutConstLabel := prometheus.NewCounterVec(
 		prometheus.CounterOpts{
 			Name: "test",
 			Help: "help",
 		},
 		[]string{"foo", "bar"},
 	)
-	equalButNotSame := prometheus.NewCounterVec(
+	equalButNotSameWithoutConstLabel := prometheus.NewCounterVec(
 		prometheus.CounterOpts{
 			Name: "test",
 			Help: "help",
 		},
 		[]string{"foo", "bar"},
 	)
 	var err error
-	if err = reg.Register(original); err != nil {
-		t.Fatal(err)
-	}
-	if err = reg.Register(equalButNotSame); err == nil {
-		t.Fatal("expected error when registering equal collector")
-	}
-	if are, ok := err.(prometheus.AlreadyRegisteredError); ok {
-		if are.ExistingCollector != original {
-			t.Error("expected original collector but got something else")
+
+	t.Run("RegisterNormallyReregisterNormally", func(t *testing.T) {
+		reg := prometheus.NewRegistry()
+		if err = reg.Register(original); err != nil {
+			t.Fatal(err)
 		}
-		if are.ExistingCollector == equalButNotSame {
-			t.Error("expected original callector but got new one")
+
+		if err = reg.Register(equalButNotSame); err == nil {
+			t.Fatal("expected error when registering equal collector")
 		}
-	} else {
-		t.Error("unexpected error:", err)
-	}
+		if are, ok := err.(prometheus.AlreadyRegisteredError); ok {
+			if are.ExistingCollector != original {
+				t.Error("expected original collector but got something else")
+			}
+			if are.ExistingCollector == equalButNotSame {
+				t.Error("expected original collector but got new one")
+			}
+		} else {
+			t.Error("unexpected error:", err)
```

Comment:
Why are you using t.Error here. I think this is never expected so we could just use fatal?

Reply:
It's definitely fatal as in "the test is broken". But I only use t.Fatal in tests if the rest of the test wouldn't make sense to run anymore. However, if the rest of the test can still in principle work and tell me about other broken things, I use t.Error so that it continue to runs.

In this case, the Error is the last thing in the (sub-)test, so the difference doesn't really matter. It's more that I personally default to t.Error for the reasons above.

Reply:
Makes sense yeah. :+1: